services/src/main/java/org/keycloak/forms/login/freemarker/AuthenticationStateCookie.java

```diff
@@ -69,7 +70,7 @@ public static void generateAndSetCookie(KeycloakSession session, RootAuthenticat
         cookie.setRemainingTabs(rootAuthSession.getAuthenticationSessions().keySet());
 
         try {
-            String encoded = JsonSerialization.writeValueAsString(cookie);
+            String encoded = Encode.urlEncode(JsonSerialization.writeValueAsString(cookie));
```

Comment:
Does it make sense to do encoding within the provider based on the cookie you are setting? For instance, the provider would decide when/how to encode a specific cookie if it requires encoding by looking at its properties from `CookieType`.

Reply:
I do actually want to do something like that, including for the JWT based cookies, but not 100% sure how that will look like and rather want to do that as a follow-up.